File: src/journal/blog.cpp
#include "journal.h"
#include "blog.h"
#include <fstream>

int create_blog(fs::path basedir, fs::path scriptdir) {
    std::string build_version = run_command("git rev-parse --short HEAD");
    std::string current_date = current_datetime();

    fs::path blog_public = basedir / "public" / "blog";
    if (!fs::exists(blog_public)) {
        fs::create_directories(blog_public);
    }

    clear_directory(blog_public);

    fs::copy(basedir / "blog", blog_public, fs::copy_options::recursive | fs::copy_options::overwrite_existing);
    fs::copy(basedir / "img" / "blog", blog_public / "img", fs::copy_options::recursive | fs::copy_options::overwrite_existing);

    // Run awk scripts
    std::string util_awk = (scriptdir / "Util.awk").string();
    std::string meta_awk = (scriptdir / "MetadataParser.awk").string();
    std::string writer_awk = (scriptdir / "MetadataWriter.awk").string();
    std::string md_file = "journal.md";

    std::string cmd = "gawk -f \"" + util_awk + "\" -f \"" + meta_awk + "\" -f \"" + writer_awk + "\" \"" + md_file + "\"";
    int ret = std::system(cmd.c_str());
    if (ret != 0) std::cerr << "MetadataWriter.awk returned " << ret << '\n';

    // Append metadata
    {
        std::ofstream metadata((basedir / "public" / "metadata.txt").string(), std::ios::app | std::ios::binary);
        metadata << "metadata_current_date " << current_date << "\n";
        metadata << "metadata_build_version " << build_version << "\n";
    }

    // Run second awk command
    std::string parser_awk = (scriptdir / "MarkdownParser.awk").string();
    std::string blogwriter_awk = (scriptdir / "BlogWriter.awk").string();
    std::string metadata_file = (basedir / "public" / "metadata.txt").string();

    cmd = "gawk -f \"" + util_awk + "\" -f \"" + parser_awk + "\" -f \"" + blogwriter_awk + "\" \"" + metadata_file + "\" \"" + md_file + "\"";
    ret = std::system(cmd.c_str());
    if (ret != 0) std::cerr << "txt2blog.awk returned " << ret << '\n';

    std::cout << "blog generated\n";

    return 0;
}

File: src/journal/dotnet/Journal/Blog.cs
using Journal.Interfaces;

namespace Journal;

public class Blog : IBlog
{
    private readonly IShell m_shell;

    public Blog(IShell shell)
    {
        m_shell = shell;
    }

    public int Create(string basedir, string scriptdir)
    {
        var scriptDir = "src";
        string blogPath = Path.Combine("public", "blog");

        Directory.SetCurrentDirectory(basedir);
        Console.WriteLine("basedir: " + basedir);

        if (Directory.Exists(blogPath))
        {
            m_shell.CleanDirectory(blogPath);
        }
        else
        {
            Directory.CreateDirectory(blogPath);
        }

        m_shell.CopyDirectory("blog", blogPath);
        m_shell.CopyDirectory(Path.Combine("img", "blog"), Path.Combine(blogPath, "img"));

        Environment.SetEnvironmentVariable("LC_ALL", "en_US.UTF-8");

        m_shell.RunProcess("gawk", $"-f {Path.Combine(scriptDir, "Util.awk")} -f {Path.Combine(scriptDir, "MetadataParser.awk")} -f {Path.Combine(scriptDir, "MetadataWriter.awk")} journal.md");

        var now = DateTimeOffset.Now;
        var offset = (now.Offset < TimeSpan.Zero ? "-" : "+") + now.Offset.ToString("hhmm");
        var currentDate = now.ToString("yyyy-MM-ddTHH:mm:ss") + offset;
        string buildVersion = m_shell.RunCommand("git", "rev-parse --short HEAD").Trim();

        using (var fs = new FileStream(Path.Combine("public", "metadata.txt"), FileMode.Append, FileAccess.Write))
        using (var writer = new StreamWriter(fs))
        {
            writer.WriteLine($"metadata_current_date {currentDate}");
            writer.WriteLine($"metadata_build_version {buildVersion}");
        }

        m_shell.RunProcess("gawk", $"-f {Path.Combine(scriptDir, "Util.awk")} -f {Path.Combine(scriptDir, "MarkdownParser.awk")} -f {Path.Combine(scriptDir, "BlogWriter.awk")} public/metadata.txt journal.md");

        Console.WriteLine("blog generated");
        return 0;
    }

}